File: C++/source/Player.cc
//
// Created by liamg on 24/10/2025.
//

#include "../header/Player.h"

#include <iostream>
#include <ostream>

#include "../header/GameBoard.h"

Player::Player(const string &name, const int startingPosition, const int startingSuggestions) {
  this->name = name;
  this->position = startingPosition;
  this->suggestions = startingSuggestions;
}

void Player::move(const int roll) {
  cout << name << " moves " << roll << " spaces" << endl;

  const int newPos = position + roll;
  int finalPos = newPos;

  if (GameBoard::isInLadders(newPos)) {
    finalPos = GameBoard::getLadderEnd(newPos);
    const int diff = finalPos - newPos;
    cout << "Hit a ladder! - moved " << diff << " spaces forwards" << endl;
  } else if (GameBoard::isInSnakes(newPos)) {
    finalPos = GameBoard::getSnakeEnd(newPos);
    const int diff = newPos - finalPos;
    cout << "Hit a snake! - moved " << diff << " spaces backwards" << endl;
  }

  this->position = finalPos;
}

File: scala/src/main/scala/Player.scala
abstract class Player(private val _name : String, private val _startingPosition : Int, private val _suggestions : Int) {
  
  protected val name: String = _name
  protected var position: Int = _startingPosition
  protected var suggestions: Int = _suggestions

  def move(roll : Int) : Unit = {
    println(name + " moves " + roll + " spaces")
    
    val newPos = position + roll
    var finalPos = newPos
    
    if (GameBoard.isInLadders(newPos)) {
      finalPos = GameBoard.getLadderEnd(newPos)
      val diff = finalPos - newPos 
      println("Hit a ladder! - moved " + diff + " spaces forwards")
    }
    else if (GameBoard.isInSnakes(newPos)) {
      finalPos = GameBoard.getSnakeEnd(newPos)
      val diff = newPos - finalPos
      println("Hit a snake! - moved " + diff + " space backwards")
    }
    
    position = finalPos
  }
  
  def handleMoveSuggestion(initialRoll : Int): Int
  
  def getName: String = name
  def getPosition : Int = position
  def getSuggestions : Int = suggestions
}
